src/optional.h

```diff
@@ -234,6 +234,30 @@ class optional
         }
 };
 
+template<class T, class U>
+constexpr bool operator==( const optional<T> &lhs, const optional<U> &rhs )
+{
+    if( bool( lhs ) != bool( rhs ) ) {
+        return false;
+    } else if( !lhs ) {
+        return true;
+    } else {
+        return *lhs == *rhs;
+    }
+}
+
+template< class T, class U >
+constexpr bool operator!=( const optional<T> &lhs, const optional<U> &rhs )
```

Comment:
`operator!=` is typically implement with a single statement: ` return !operator==( lhs, rhs );` - this ensures the relation between `==` and `!=` is as one would expect them.